homeassistant/components/govee_light_api/coordinator.py

```diff
@@ -0,0 +1,122 @@
+"""Coordinator for Govee Local API."""
+
+from collections.abc import Callable
+from datetime import timedelta
+import logging
+
+from govee_local_api import GoveeController, GoveeDevice
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    CONF_BIND_ADDRESS,
+    CONF_DISCOVERY_INTERVAL,
+    CONF_DISCOVERY_INTERVAL_DEFAULT,
+    CONF_LISENING_PORT,
+    CONF_MULTICAST_ADDRESS,
+    CONF_TARGET_PORT,
+)
+
+
+class GoveeLocalApiCoordinator(DataUpdateCoordinator):
+    """My custom coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        scan_interval: timedelta,
+        logger: logging.Logger,
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass=hass,
+            logger=logger,
+            name="GoveeLightLocalApi",
+            update_interval=scan_interval,
+        )
+
+        self._config_entry = config_entry
+        config = config_entry.data["config"]
+
+        option_discovery_interval = config_entry.options.get(CONF_DISCOVERY_INTERVAL)
+        self._discovery_interval = (
+            option_discovery_interval
+            if option_discovery_interval
+            else config.get(CONF_DISCOVERY_INTERVAL, CONF_DISCOVERY_INTERVAL_DEFAULT)
+        )
+
+        self._controller = GoveeController(
+            loop=hass.loop,
+            logger=logger,
+            listening_address=config[CONF_BIND_ADDRESS],
+            broadcast_address=config[CONF_MULTICAST_ADDRESS],
+            broadcast_port=config[CONF_TARGET_PORT],
+            listening_port=config[CONF_LISENING_PORT],
+            discovery_enabled=True,
+            discovery_interval=config[CONF_DISCOVERY_INTERVAL],
+            discovered_callback=None,
+            update_enabled=False,
+        )
+
+    async def start(self) -> None:
+        """Start the Govee coordinator."""
+        await self._controller.start()
+        self._controller.send_update_message()
+
+    async def set_discovery_callback(
+        self, callback: Callable[[GoveeDevice, bool], bool]
+    ) -> None:
+        """Set discovery callback for automatic Govee light discovery."""
+        self._controller.set_device_discovered_callback(callback)
+
+    def clenaup(self) -> None:
+        """Stop and clenaup the cooridinator."""
+        self._controller.clenaup()
+
+    async def turn_on(self, device: GoveeDevice) -> None:
+        """Turn on the light."""
+        assert self._controller == device.controller
```

Comment:
why do we do this?

Reply:
This is an excellent question 😅 I remember writing this in an early draft of the code, but it makes little sense now